src/Language/PureScript/Sugar/TypeClasses/Instances.hs

```diff
@@ -0,0 +1,63 @@
+-- |
+-- This module implements the desugaring pass which creates the compiler-generated
+-- names for type class instances that do not have one defined in the source code.
+--
+module Language.PureScript.Sugar.TypeClasses.Instances
+  ( desugarTypeClassInstanceNames
+  ) where
+
+import Prelude.Compat hiding (take)
+
+import           Control.Monad.Error.Class (MonadError(..))
+import           Control.Monad.Supply.Class
+import qualified Data.Map as M
+import           Data.Maybe (catMaybes)
+import           Data.Text (Text, take, pack)
+import           Language.PureScript.Errors
+import           Language.PureScript.Names
+import           Language.PureScript.TypeChecker.Monad
+
+-- |
+-- Completes the name generation for type class instances that do not have
+-- a unique name defined in source code. All `Left Text` values
+-- will be converted to `Right Ident` values.
+--
+desugarTypeClassInstanceNames
+  :: (MonadSupply m, MonadError MultipleErrors m)
+  => Module
+  -> m (M.Map Text Ident, Module)
```

Comment:
On review, I'm not so much a fan of this instance map thing. Maybe I'm too attached to my idea of completely replacing the chain ID type with some other unique identifier, but the comments you left on your investigation into the chain ID seemed to indicate that it could work, and it'd solve the duplicate generated name problem and keep the desugaring pipeline neater. Did you find something that made you think it wouldn't work? If not, I could take a swing at doing that refactoring in my own PR if you'd rather not undertake it here.